blog/2024-03-27-reductstore-vs-timescale/index.md

```diff
@@ -0,0 +1,220 @@
+    ) as reduct_client:
+        count = 0
+        bucket = await reduct_client.get_bucket("benchmark")
+        for i in range(1, BLOB_COUNT):
+            await bucket.write("data", CHUNK)
+            count += BLOB_SIZE
+        return count
+
+async def read_from_reduct(t1, t2):
+    async with ReductClient(
+        f"http://{HOST}:8383", api_token="reductstore"
+    ) as reduct_client:
+        count = 0
+        bucket = await reduct_client.get_bucket("benchmark")
+        async for rec in bucket.query("data", t1, t2):
+            count += len(await rec.read_all())
+        return count
+```
+
+## Benchmarks[](https://www.reduct.store/blog/comparisons/computer-vision/iot/performance-comparison-reductstore-vs-minio#benchmarks)
+
+After establishing our read/write functions, we can start writing our benchmarks.
+
+```python
+if __name__ == "__main__":
+    print(f"Chunk size={BLOB_SIZE / 1000_000} Mb, count={BLOB_COUNT}")
+    ts = time.time()
+    size = write_to_timescale()
+    print(f"Write {size / 1000_000} Mb to TimescaleDB: {time.time() - ts} s")
+
+    ts_read = time.time()
+    size = read_from_timescale(ts, time.time())
+    print(f"Read {size / 1000_000} Mb from TimescaleDB: {time.time() - ts_read} s")
+
+    loop = asyncio.new_event_loop()
+    ts = time.time()
+    size = loop.run_until_complete(write_to_reduct())
+    print(f"Write {size / 1000_000} Mb to ReductStore: {time.time() - ts} s")
+
+    ts_read = time.time()
+    size = loop.run_until_complete(read_from_reduct(ts, time.time()))
+    print(f"Read {size / 1000_000} Mb from ReductStore: {time.time() - ts_read} s")
+```
+
+For testing purposes, we need to run the databases. This can easily be done using docker-compose:
+
+```yaml
+version: "3"
+services:
+  timescale:
+    image: timescale/timescaledb:latest-pg14
+    ports:
+      - "5432:5432"
+    environment:
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: postgres
+    volumes:
+      - ${PWD}/data/timescale:/var/lib/postgresql/data
+
+  reductstore:
+    image: reduct/store:latest
+    ports:
+      - "8383:8383"
+    environment:
+      RS_API_TOKEN: reductstore
+      RS_BUCKET_1_NAME: benchmark
+      RS_BUCKET_1_QUTA_TYPE: FIFO
+      RS_BUCKET_1_QUOTA_SIZE: 50TB
+
+    volumes:
+      - ${PWD}/data/reductstore:/data
+```
+
+Take note, ReductStore enables the provisioning of any resources with environment variables. In this case, we create a bucket called `benchmark` with a 50TB quota. Your DevOps team will appreciate it.
+
+Let's set up and run benchmarks:
+
+```yaml
+docker-compose up -d
+python3 main.py
+```
+
+## Results
+
+The script displays results for the specified `BLOB_SIZE` and `SIZE_COUNT`. On my device, which has an NVMe drive, here are the results I obtained:
+
+| Chunk Size | Operation | TimescaleDB, blob/s | ReductStore, blob/s |
+| --- | --- | --- | --- |
+| 10 KB | Write | 1557 | 1500 |
+|  | Read | 1333 | 1280 |
+| 100 KB | Write | 447 | 1366 |
```

Comment:
I see that it significantly outperforms timescale for blobs greater than 10KB but it can be more highlighted.

If you add a percentage in the table to compare the speed differences or maybe an extra example with a typical scenario and compare the time needed to store and retrieve lot of data to show how compounding extra read time can be extremely frustrating for TimescaleDB users.

Reply:
## Other Considerations

When we choose a database for storing blob data, we should consider not only performance but also the following factors:

* **Retention Policy**: ReductStore provides a retention policy based on disk usage, which is crucial when the amount of data stored over time is unpredictable.
* **Data Size**: ReductStore is more efficient for storing large blobs, while TimescaleDB is better for small blobs.
* **Querying**: TimescaleDB is better for structured data and complex queries, while ReductStore is more suitable for storing and retrieving blobs with minimal overhead and latency.
* **Replication**: TimescaleDB replicates data across multiple nodes as exact copies, while ReductStore provides append-only replication, with the ability to filter data based on labels. This could be a part of your data reduction strategy.

## Conclusions

TimescaleDB excels as a time series database and can be used to store small blobs. It's a good choice if you need to store structured data and want to avoid additional storage for blob data. However, if you have blobs larger than 10KB, ReductStore might be a better option. It offers better performance and provides a retention policy based on disk usage and conditional append-only replication for your data reduction strategy.